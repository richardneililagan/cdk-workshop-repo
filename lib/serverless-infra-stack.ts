import * as cdk from '@aws-cdk/core';
import * as nodejs from '@aws-cdk/aws-lambda-nodejs'
import * as ddb from '@aws-cdk/aws-dynamodb'
import * as iam from '@aws-cdk/aws-iam'

import { DDBTableWriter } from './constructs/ddb-table-writer'

// :: ---

export interface ServerlessInfraStackProps extends cdk.StackProps {
  //
}

export class ServerlessInfraStack extends cdk.Stack {
  // :: ---
  // :: scope === parent
  constructor(scope: cdk.Construct, id: string, props?: ServerlessInfraStackProps) {
    super(scope, id, props);
    
    // :: our DDB table
    const table = new ddb.Table(this, 'my-ddb-table', {
      partitionKey: {
        name: 'new-id',
        type: ddb.AttributeType.STRING
      }
    })
    
    new DDBTableWriter(this, 'my-ddb-table-writer', {
      table: table
    })
  }
}
