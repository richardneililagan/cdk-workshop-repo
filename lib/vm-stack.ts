import * as cdk from '@aws-cdk/core'
import * as ec2 from '@aws-cdk/aws-ec2'

// :: ---

export interface VmStackProps extends cdk.StackProps {
  //
}

export class VmStack extends cdk.Stack {
  vpc: ec2.Vpc
  
  constructor (scope: cdk.Construct, id: string, props?: VmStackProps) {
    super(scope, id, props)
  
    // :: create a new network
    this.vpc = new ec2.Vpc(this, 'vm-network', {
      maxAzs: 3,
      cidr: '50.0.0.0/16'
    })
    
    // :: create a new VM / EC2
    new ec2.Instance(this, 'my-vm', {
      vpc: this.vpc,
      instanceType: new ec2.InstanceType('m5.large'),
      machineImage: ec2.MachineImage.latestAmazonLinux()
    })
  }
  
  onValidate(): string[] {
    return []
  }
}