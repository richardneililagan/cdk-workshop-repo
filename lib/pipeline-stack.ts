import * as cdk from '@aws-cdk/core'

import { CdkPipeline, SimpleSynthAction } from '@aws-cdk/pipelines'
import * as codepipeline from '@aws-cdk/aws-codepipeline'
import * as actions from '@aws-cdk/aws-codepipeline-actions'

import { ServerlessInfraStack } from './serverless-infra-stack';
import { VmStack } from './vm-stack'

// :: ---

class Application extends cdk.Stage {
    constructor (scope: cdk.Construct, id: string, props?: cdk.StageProps) {
        super(scope, id, props)
        
        new ServerlessInfraStack(scope, 'ServerlessInfraStack');
        new VmStack(scope, 'vm-stack')
    }
}

export class PipelineStack extends cdk.Stack {
    constructor (scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props)
        
        // :: (1) define trigger
        // :: (2) define build + compile
        // :: (3) define the deploy
        
        const sourceArtifact = new codepipeline.Artifact()
        const cloudAssemblyArtifact = new codepipeline.Artifact()
        
        // :: (1) --- this triggers when code changes in the repo
        const sourceAction = new actions.GithubSourceAction({
            actionName: 'new-code-change-trigger',
            output: sourceArtifact,
            owner: 'richardneililagan',
            repo: 'cdk-workshop-repo',
            branch: 'master',
            oauthToken: cdk.SecretValue.secretsManager('workshop/cdk/github-token')
        })
        
        // :: (2) --- 
        const synthAction = SimpleSynthAction.standardNpmSynth({
            sourceArtifact: sourceArtifact,
            cloudAssemblyArtifact: cloudAssemblyArtifact,
            environment: { privileged: true }
        })
        
        // :: (3) ---
        const pipeline = new CdkPipeline(this, 'app-pipeline', {
            cloudAssemblyArtifact: cloudAssemblyArtifact,
            sourceAction: sourceAction,
            synthAction: synthAction
        })
        
        pipeline.addApplicationStage(new Application(this, 'production-environment'))
    }
}