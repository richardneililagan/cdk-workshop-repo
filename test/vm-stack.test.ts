import * as cdk from '@aws-cdk/core'
import * as ec2 from '@aws-cdk/aws-ec2'
import { SynthUtils } from '@aws-cdk/assert'
import '@aws-cdk/assert/jest'

import { VmStack } from '../lib/vm-stack'

// :: ---

// :: TODO smoke test
it('synthesizes without errors', () => {
    const app = new cdk.App()
    const stack = new VmStack(app, 'my-vm-stack')
    
    const template = SynthUtils.toCloudFormation(stack)
})

it('has a network using CIDR 50.0.0.0/16', () => {
    const app = new cdk.App()
    const stack = new VmStack(app, 'my-vm-stack')
    
    expect(stack).toHaveResource('AWS::EC2::VPC', {
        CidrBlock: '50.0.0.0/16'
    })
})

it('has all instances running Linux', () => {
    const app = new cdk.App()
    const stack = new VmStack(app, 'my-vm-stack')
    
    const instances = stack.node.children.filter(
        child => child instanceof ec2.Instance
    )
    
    for (const instance of instances) {
        const __instance = instance as ec2.Instance
        
        expect(__instance.osType).toBe(ec2.OperatingSystemType.LINUX)
    }
})