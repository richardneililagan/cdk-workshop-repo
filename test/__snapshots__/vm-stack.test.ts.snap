// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`synthesizes without errors 1`] = `
Object {
  "Resources": Object {
    "myvmC57D50F0": Object {
      "DependsOn": Array [
        "myvmInstanceRole95C64C40",
      ],
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "IamInstanceProfile": Object {
          "Ref": "myvmInstanceProfile22D0C72D",
        },
        "ImageId": "{{resolve:ssm:/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2}}",
        "InstanceType": "m5.large",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "myvmInstanceSecurityGroup43ACC3CF",
              "GroupId",
            ],
          },
        ],
        "SubnetId": Object {
          "Ref": "vmnetworkPrivateSubnet1Subnet8B8A5826",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/my-vm",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "myvmInstanceProfile22D0C72D": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "myvmInstanceRole95C64C40",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "myvmInstanceRole95C64C40": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/my-vm",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "myvmInstanceSecurityGroup43ACC3CF": Object {
      "Properties": Object {
        "GroupDescription": "my-vm-stack/my-vm/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/my-vm",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "vmnetworkE26026A1": Object {
      "Properties": Object {
        "CidrBlock": "50.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vmnetworkIGW3A1CD95F": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "vmnetworkPrivateSubnet1DefaultRoute8716CA56": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "vmnetworkPublicSubnet1NATGatewayF00774EF",
        },
        "RouteTableId": Object {
          "Ref": "vmnetworkPrivateSubnet1RouteTable6FF830C4",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vmnetworkPrivateSubnet1RouteTable6FF830C4": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vmnetworkPrivateSubnet1RouteTableAssociation344AF5DD": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vmnetworkPrivateSubnet1RouteTable6FF830C4",
        },
        "SubnetId": Object {
          "Ref": "vmnetworkPrivateSubnet1Subnet8B8A5826",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vmnetworkPrivateSubnet1Subnet8B8A5826": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "50.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vmnetworkPrivateSubnet2DefaultRoute26FE5B98": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "vmnetworkPublicSubnet2NATGateway65E21873",
        },
        "RouteTableId": Object {
          "Ref": "vmnetworkPrivateSubnet2RouteTable8E5F3CBE",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vmnetworkPrivateSubnet2RouteTable8E5F3CBE": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vmnetworkPrivateSubnet2RouteTableAssociation4C23B956": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vmnetworkPrivateSubnet2RouteTable8E5F3CBE",
        },
        "SubnetId": Object {
          "Ref": "vmnetworkPrivateSubnet2SubnetE8178A7D",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vmnetworkPrivateSubnet2SubnetE8178A7D": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "50.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vmnetworkPublicSubnet1DefaultRouteC4C5DFE2": Object {
      "DependsOn": Array [
        "vmnetworkVPCGW89A55A4F",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vmnetworkIGW3A1CD95F",
        },
        "RouteTableId": Object {
          "Ref": "vmnetworkPublicSubnet1RouteTable8C41811C",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vmnetworkPublicSubnet1EIPE8CF5578": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vmnetworkPublicSubnet1NATGatewayF00774EF": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "vmnetworkPublicSubnet1EIPE8CF5578",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "vmnetworkPublicSubnet1SubnetDCD9367F",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vmnetworkPublicSubnet1RouteTable8C41811C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vmnetworkPublicSubnet1RouteTableAssociation20A4E0DD": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vmnetworkPublicSubnet1RouteTable8C41811C",
        },
        "SubnetId": Object {
          "Ref": "vmnetworkPublicSubnet1SubnetDCD9367F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vmnetworkPublicSubnet1SubnetDCD9367F": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "50.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vmnetworkPublicSubnet2DefaultRouteB2EDDA9A": Object {
      "DependsOn": Array [
        "vmnetworkVPCGW89A55A4F",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vmnetworkIGW3A1CD95F",
        },
        "RouteTableId": Object {
          "Ref": "vmnetworkPublicSubnet2RouteTableBE56B8EB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vmnetworkPublicSubnet2EIP923C9973": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vmnetworkPublicSubnet2NATGateway65E21873": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "vmnetworkPublicSubnet2EIP923C9973",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "vmnetworkPublicSubnet2Subnet1E0F71C5",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vmnetworkPublicSubnet2RouteTableAssociationCDFDA06C": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vmnetworkPublicSubnet2RouteTableBE56B8EB",
        },
        "SubnetId": Object {
          "Ref": "vmnetworkPublicSubnet2Subnet1E0F71C5",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vmnetworkPublicSubnet2RouteTableBE56B8EB": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vmnetworkPublicSubnet2Subnet1E0F71C5": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "50.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "my-vm-stack/vm-network/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vmnetworkVPCGW89A55A4F": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "vmnetworkIGW3A1CD95F",
        },
        "VpcId": Object {
          "Ref": "vmnetworkE26026A1",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
